apps/emqx/src/emqx_persistent_message_ds_replayer.erl

```diff
@@ -75,59 +97,81 @@ next_packet_id(Inflight0 = #inflight{next_seqno = LastSeqno}) ->
     {Replies, Inflight}.
 
--spec commit_offset(emqx_persistent_session_ds:id(), emqx_types:packet_id(), inflight()) ->
+-spec commit_offset(emqx_persistent_session_ds:id(), track(), emqx_types:packet_id(), inflight()) ->
     {_IsValidOffset :: boolean(), inflight()}.
 commit_offset(
     SessionId,
+    Track,
     PacketId,
-    Inflight0 = #inflight{
-        acked_until = AckedUntil, next_seqno = NextSeqno
-    }
+    Inflight0 = #inflight{commits = Commits}
 ) ->
-    case packet_id_to_seqno(NextSeqno, PacketId) of
-        Seqno when Seqno >= AckedUntil andalso Seqno < NextSeqno ->
+    case validate_commit(Track, PacketId, Inflight0) of
+        CommitUntil when is_integer(CommitUntil) ->
             %% TODO
-            %% We do not preserve `acked_until` in the database. Instead, we discard
+            %% We do not preserve `CommitUntil` in the database. Instead, we discard
             %% fully acked ranges from the database. In effect, this means that the
-            %% most recent `acked_until` the client has sent may be lost in case of a
+            %% most recent `CommitUntil` the client has sent may be lost in case of a
             %% crash or client loss.
-            Inflight1 = Inflight0#inflight{acked_until = next_seqno(Seqno)},
-            Inflight = discard_acked(SessionId, Inflight1),
+            Inflight1 = Inflight0#inflight{commits = Commits#{Track := CommitUntil}},
+            Inflight = discard_committed(SessionId, Inflight1),
             {true, Inflight};
-        OutOfRange ->
-            ?SLOG(warning, #{
-                msg => "out-of-order_ack",
-                acked_until => AckedUntil,
-                acked_seqno => OutOfRange,
-                next_seqno => NextSeqno,
-                packet_id => PacketId
-            }),
+        false ->
+            {false, Inflight0}
+    end.
+
+-spec commit_marker(emqx_persistent_session_ds:id(), marker(), emqx_types:packet_id(), inflight()) ->
+    {_IsValidMarker :: boolean(), inflight()}.
+commit_marker(
+    SessionId,
+    Marker = rec,
+    PacketId,
+    Inflight0 = #inflight{commits = Commits}
+) ->
+    case validate_commit(Marker, PacketId, Inflight0) of
+        CommitUntil when is_integer(CommitUntil) ->
+            update_marker(SessionId, Marker, CommitUntil),
+            Inflight = Inflight0#inflight{commits = Commits#{Marker := CommitUntil}},
+            {true, Inflight};
+        false ->
             {false, Inflight0}
     end.
 
--spec poll(emqx_persistent_session_ds:id(), inflight(), pos_integer()) ->
+-spec poll(reply_fun(), emqx_persistent_session_ds:id(), inflight(), pos_integer()) ->
     {emqx_session:replies(), inflight()}.
-poll(SessionId, Inflight0, WindowSize) when WindowSize > 0, WindowSize < 16#7fff ->
-    #inflight{next_seqno = NextSeqNo0, acked_until = AckedSeqno} =
-        Inflight0,
+poll(ReplyFun, SessionId, Inflight0, WindowSize) when WindowSize > 0, WindowSize < ?EPOCH_SIZE ->
```

Comment:
```suggestion
poll(ReplyFun, SessionId, Inflight0, WindowSize) when WindowSize > 0, WindowSize =< ?EPOCH_SIZE ->
```

?

Reply:
> Followed by the Two Byte Integer representing the Receive Maximum value.

The `?EPOCH_SIZE` is defined as `16#10000` (same constant that's used for the purpose of skipping seqnos), so it currently seems correct?

Reply:
Right, sorry.